gpcontrib/arenadata_toolkit/src/arenadata_toolkit_worker.c

```diff
@@ -0,0 +1,361 @@
+#include "postgres.h"
+
+#include "access/xact.h"
+#include "access/genam.h"
+#include "access/heapam.h"
+#include "catalog/pg_db_role_setting.h"
+#include "catalog/pg_extension.h"
+#include "catalog/indexing.h"
+#include "cdb/cdbdisp_query.h"
+#include "cdb/cdbdispatchresult.h"
+#include "cdb/cdbvars.h"
+#include "libpq-fe.h"
+#include "postmaster/bgworker.h"
+#include "storage/proc.h"
+#include "storage/ipc.h"
+#include "utils/snapmgr.h"
+#include "utils/builtins.h"
+#include "utils/fmgroids.h"
+
+#include "arenadata_toolkit_worker.h"
+#include "arenadata_toolkit_guc.h"
+#include "bloom_set.h"
+#include "tf_shmem.h"
+
+#define TOOLKIT_BINARY_NAME "arenadata_toolkit"
+
+typedef struct
+{
+	Oid			dbid;
+	Name		dbname;
+	bool		get_full_snapshot_on_recovery;
+}	tracked_db_t;
+
+/* flags set by signal handlers */
+static volatile sig_atomic_t got_sighup = false;
+static volatile sig_atomic_t got_sigterm = false;
+
+void arenadata_toolkit_main(Datum);
+
+/* parse array of GUCs, find desired and analyze it */
+static bool
+is_db_tracked(ArrayType *array)
+{
+	bool		is_tracked = false;
+	Datum	   *elems;
+	bool	   *nulls;
+	int			nelems;
+
+	deconstruct_array(array, TEXTOID, -1, false, 'i',
+					  &elems, &nulls, &nelems);
+	for (int i = 0; i < nelems; i++)
+	{
+		char	   *s;
+		char	   *name;
+		char	   *value;
+
+		if (nulls[i])
+			continue;
+
+		s = TextDatumGetCString(elems[i]);
+		ParseLongOption(s, &name, &value);
+
+		if (!value)
+		{
+			free(name);
+			continue;
+		}
+
+		if (strcmp(name, "arenadata_toolkit.tracking_is_db_tracked") == 0 &&
+			strcmp(value, "t") == 0)
+		{
+			is_tracked = true;
+			break;
+		}
+
+		free(name);
+		if (value)
+			free(value);
+		pfree(s);
+	}
+
+	return is_tracked;
+}
+
+static bool
+full_snapshot_on_recovery(ArrayType *array)
```

Comment:
The functions is_db_tracked and full_snapshot_on_recovery are very similar internally. Can you extract the common parts into a separate function?